### PR TITLE
NPC - Boss features

### PR DESCRIPTION
**Pull request by MikeJeffers**
Closes: https://github.com/OpenPerpetuum/OP-Project/issues/138
Requires: https://github.com/OpenPerpetuum/OPDB/pull/134
NpcSpecialType enum added to indicate flocks of a "boss" type configuration.
Boss features:
 - Automated loot splitting to all participants
 - Random shuffle on respawn timer (+/-10%)

### CODE REVIEW

File: src/Perpetuum/Services/Looting/LootGenerator.cs [hunk 1]
```diff
@@ -25,4 +34,68 @@ public IEnumerable<LootItem> Generate()
...
+                for (var i = 0; i < splitCount; i++)
+                {
+                    var amount = splitQuantity + (remainder > 0 ? 1 : 0);
+                    remainder--;
+                    amounts[i] = amount;
+                }
+                amounts = amounts.OrderBy(x => _random.Next()).ToArray();
+                for (var i = 0; i < splitCount; i++)
+                {
+                    ItemInfo itemInfo = new ItemInfo(lootItem.ItemInfo.Definition, lootItem.Quantity, lootItem.Quantity);
```
Comment (MikeJeffers) [inline src/Perpetuum/Services/Looting/LootGenerator.cs]: This should be splitquantity plz fix

File: src/Perpetuum/Services/Looting/CompositeLootGenerator.cs [hunk 2]
```diff
@@ -11,6 +12,20 @@ public CompositeLootGenerator(params ILootGenerator[] generators)
             _generators = generators;
         }
 
+        public IEnumerable<LootGeneratorItemInfo> GetInfos()
+        {
+            var infos = _generators.Select(x => x.GetInfos());
+            var allInfos = new List<LootGeneratorItemInfo>();
+            foreach (var infoset in infos)
```
Comment (nref-dan) [inline src/Perpetuum/Services/Looting/CompositeLootGenerator.cs]: If I'm reading this right, `_generators.Select(...)` generates an `IEnumerable<IEnumerable<LootGeneratorItemInfo>>`
In which case you can shorten it to this:
`infos.SelectMany(infoset => infoset).ToList();`

Also, as with the relic code, be careful passing IEnumerable around. You're not really passing the data, just a query pointer for deferred execution, the garbage collector will get really confused trying to keep these classes alive due to external pointing of a locally scoped object.

**Review by nref-dan — CHANGES_REQUESTED** [review]

File: src/Perpetuum/Services/Looting/CompositeLootGenerator.cs (hunk 2 above)
Comment (MikeJeffers) [reply src/Perpetuum/Services/Looting/CompositeLootGenerator.cs]: With the .toList finalizer on the return, while it is returned as a IEnumerable, the promise of having data will have been fulfilled - is that a fair characterization?

Change as requested incoming

**Review by nref-dan — CHANGES_REQUESTED** [review]
Where you've changed the signature to return an IReadOnlyCollection make sure you call `AsReadOnly()` on the list before you return it. This will prevent the caller from being able to cast it to a modifiable list.

**Review by nref-dan — APPROVED** [review]

**Review by clouths — APPROVED** [review]
LGTM

**Pull request merged**